<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Kwon Lab Pipelines - Searching for genes in bacterial genomes</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../">
<link href="../../../materials/metagenomics/virgo2/index.html" rel="prev">
<script src="../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<link href="../../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../index.html">
    <span class="navbar-title">Kwon Lab Pipelines</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://kwonlab.mgh.harvard.edu/"> 
<span class="menu-text">About the Kwon Lab</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/kwondry/documentation"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <a class="flex-grow-1 no-decor" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
          <h1 class="quarto-secondary-nav-title">Searching for genes in bacterial genomes</h1>
        </a>     
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Getting Started</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/getting-started/o2/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"> Get your O2 account</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/getting-started/installation/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"> Installing software for work on O2</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">R and R Studio</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/getting-started/local_R/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"> RStudio and recommended packages</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/r-tutorial/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"> R workshop</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">RNAseq</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/rnaseq/background/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Background</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/rnaseq/walkthrough/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Walkthrough</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">16S</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/16S/background/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Background</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/16S/kwon_sequencing_data_transfer/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">SMB Data Transfer and Demultiplexing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/16S/16s_pipeline/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">16S Pipeline</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/16S/valencia/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assigning CSTs with Valencia</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Metagenomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/metagenomics/background/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Background</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/metagenomics/virgo2/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">VIRGO2 Pipeline</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
 <span class="menu-text">Comparative Genomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../materials/comparative-genomics/gene-search/index.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Gene Search</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#preparing-your-reference-database" id="toc-preparing-your-reference-database" class="nav-link active" data-scroll-target="#preparing-your-reference-database">Preparing Your Reference Database</a>
  <ul class="collapse">
  <li><a href="#combine-genome-protein-files" id="toc-combine-genome-protein-files" class="nav-link" data-scroll-target="#combine-genome-protein-files">Combine Genome Protein Files</a></li>
  <li><a href="#alternative-combine-with-custom-glob-pattern" id="toc-alternative-combine-with-custom-glob-pattern" class="nav-link" data-scroll-target="#alternative-combine-with-custom-glob-pattern">Alternative: Combine with Custom Glob Pattern</a></li>
  </ul></li>
  <li><a href="#getting-your-query-gene" id="toc-getting-your-query-gene" class="nav-link" data-scroll-target="#getting-your-query-gene">Getting Your Query Gene</a>
  <ul class="collapse">
  <li><a href="#step-1-find-your-taxon-and-annotation" id="toc-step-1-find-your-taxon-and-annotation" class="nav-link" data-scroll-target="#step-1-find-your-taxon-and-annotation">Step 1: Find Your Taxon and Annotation</a></li>
  <li><a href="#step-2-decide-on-search-strategy" id="toc-step-2-decide-on-search-strategy" class="nav-link" data-scroll-target="#step-2-decide-on-search-strategy">Step 2: Decide on Search Strategy</a></li>
  </ul></li>
  <li><a href="#specific-search-approach-sequence-based" id="toc-specific-search-approach-sequence-based" class="nav-link" data-scroll-target="#specific-search-approach-sequence-based">Specific Search Approach (Sequence-based)</a>
  <ul class="collapse">
  <li><a href="#get-the-uniref90-representative-sequence" id="toc-get-the-uniref90-representative-sequence" class="nav-link" data-scroll-target="#get-the-uniref90-representative-sequence">1. Get the UniRef90 Representative Sequence</a></li>
  <li><a href="#search-against-genome-database-using-mmseqs2" id="toc-search-against-genome-database-using-mmseqs2" class="nav-link" data-scroll-target="#search-against-genome-database-using-mmseqs2">2. Search Against Genome Database Using MMseqs2</a></li>
  </ul></li>
  <li><a href="#domain-based-search-approach-using-ncbi-cog-pssms" id="toc-domain-based-search-approach-using-ncbi-cog-pssms" class="nav-link" data-scroll-target="#domain-based-search-approach-using-ncbi-cog-pssms">Domain-based Search Approach (Using NCBI COG PSSMs)</a>
  <ul class="collapse">
  <li><a href="#download-the-cog-database" id="toc-download-the-cog-database" class="nav-link" data-scroll-target="#download-the-cog-database">1. Download the COG Database</a></li>
  <li><a href="#identify-your-cog-of-interest" id="toc-identify-your-cog-of-interest" class="nav-link" data-scroll-target="#identify-your-cog-of-interest">2. Identify Your COG of Interest</a></li>
  <li><a href="#search-against-genome-database-using-rps-blast" id="toc-search-against-genome-database-using-rps-blast" class="nav-link" data-scroll-target="#search-against-genome-database-using-rps-blast">3. Search Against Genome Database Using RPS-BLAST</a></li>
  <li><a href="#filter-for-your-specific-cog" id="toc-filter-for-your-specific-cog" class="nav-link" data-scroll-target="#filter-for-your-specific-cog">4. Filter for Your Specific COG</a></li>
  </ul></li>
  <li><a href="#broad-search-approach-hmm-based" id="toc-broad-search-approach-hmm-based" class="nav-link" data-scroll-target="#broad-search-approach-hmm-based">Broad Search Approach (HMM-based)</a>
  <ul class="collapse">
  <li><a href="#get-query-sequences-from-uniref50-cluster" id="toc-get-query-sequences-from-uniref50-cluster" class="nav-link" data-scroll-target="#get-query-sequences-from-uniref50-cluster">1. Get Query Sequences from UniRef50 Cluster</a></li>
  <li><a href="#align-the-sequences" id="toc-align-the-sequences" class="nav-link" data-scroll-target="#align-the-sequences">2. Align the Sequences</a></li>
  <li><a href="#build-the-hmm-profile" id="toc-build-the-hmm-profile" class="nav-link" data-scroll-target="#build-the-hmm-profile">3. Build the HMM Profile</a></li>
  <li><a href="#search-against-genome-database-using-hmmsearch" id="toc-search-against-genome-database-using-hmmsearch" class="nav-link" data-scroll-target="#search-against-genome-database-using-hmmsearch">4. Search Against Genome Database Using hmmsearch</a></li>
  </ul></li>
  <li><a href="#example-downstream-analysis" id="toc-example-downstream-analysis" class="nav-link" data-scroll-target="#example-downstream-analysis">Example Downstream Analysis</a>
  <ul class="collapse">
  <li><a href="#extract-hit-ids" id="toc-extract-hit-ids" class="nav-link" data-scroll-target="#extract-hit-ids">1. Extract Hit IDs</a></li>
  <li><a href="#retrieve-hit-sequences" id="toc-retrieve-hit-sequences" class="nav-link" data-scroll-target="#retrieve-hit-sequences">2. Retrieve Hit Sequences</a></li>
  <li><a href="#align-hit-sequencess" id="toc-align-hit-sequencess" class="nav-link" data-scroll-target="#align-hit-sequencess">3. Align Hit Sequencess</a></li>
  <li><a href="#build-phylogenetic-tree" id="toc-build-phylogenetic-tree" class="nav-link" data-scroll-target="#build-phylogenetic-tree">4. Build Phylogenetic Tree</a></li>
  </ul></li>
  <li><a href="#notes" id="toc-notes" class="nav-link" data-scroll-target="#notes">Notes</a>
  <ul class="collapse">
  <li><a href="#e-value-thresholds" id="toc-e-value-thresholds" class="nav-link" data-scroll-target="#e-value-thresholds">E-value Thresholds</a></li>
  <li><a href="#performance-tips" id="toc-performance-tips" class="nav-link" data-scroll-target="#performance-tips">Performance Tips</a></li>
  <li><a href="#cog-database-updates" id="toc-cog-database-updates" class="nav-link" data-scroll-target="#cog-database-updates">COG Database Updates</a></li>
  <li><a href="#phylogenetic-analysis" id="toc-phylogenetic-analysis" class="nav-link" data-scroll-target="#phylogenetic-analysis">Phylogenetic Analysis</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">Searching for genes in bacterial genomes</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Do you have a gene (or several) that you want to get out of a genome (or several)? This guide will explain some common ways you can search genomes for genes.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Preflight checks
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<section id="installing-mamba" class="level3">
<h3 class="anchored" data-anchor-id="installing-mamba">Installing Mamba</h3>
<p>Mamba is a fast, drop-in replacement for conda that significantly speeds up package installation and dependency resolution.</p>
<p>Mamba is a fast, drop-in replacement for conda that significantly speeds up package installation and dependency resolution.</p>
<p>Type <code>which mamba</code> to check if you have mamba installed. If not, check if you have conda with <code>which conda</code></p>
<section id="install-mamba-into-existing-conda" class="level4">
<h4 class="anchored" data-anchor-id="install-mamba-into-existing-conda">Install Mamba into existing conda</h4>
<p>If you already have conda installed:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-n</span> base <span class="at">-c</span> conda-forge mamba</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="if-no-conda---install-miniforge-for-scratch" class="level4">
<h4 class="anchored" data-anchor-id="if-no-conda---install-miniforge-for-scratch">If no conda - Install Miniforge for scratch</h4>
<p>Miniforge comes with mamba pre-installed:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download and install Miniforge</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="at">-L</span> <span class="at">-O</span> <span class="st">"https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-</span><span class="va">$(</span><span class="fu">uname</span><span class="va">)</span><span class="st">-</span><span class="va">$(</span><span class="fu">uname</span> <span class="at">-m</span><span class="va">)</span><span class="st">.sh"</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">bash</span> Miniforge3-<span class="va">$(</span><span class="fu">uname</span><span class="va">)</span>-<span class="va">$(</span><span class="fu">uname</span> <span class="at">-m</span><span class="va">)</span>.sh</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>After installing Miniforge, you should type <code>conda init</code> and then log out and back into the shell.</p>
</section>
</section>
<section id="installing-required-bioinformatics-tools" class="level3">
<h3 class="anchored" data-anchor-id="installing-required-bioinformatics-tools">Installing Required Bioinformatics Tools</h3>
<p>Once you are sure you have mamba installed (<code>which mamba</code>), install the necessary tools for this searching.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install all tools in one command</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">mamba</span> create <span class="at">--name</span><span class="op">=</span>gene_search <span class="at">-c</span> conda-forge <span class="at">-c</span> bioconda mmseqs2 hmmer seqkit mafft fasttree blast</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="verify-installation" class="level4">
<h4 class="anchored" data-anchor-id="verify-installation">Verify Installation</h4>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate gene_search</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">mmseqs</span> version</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="ex">hmmbuild</span> <span class="at">-h</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="ex">seqkit</span> version</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="ex">mafft</span> <span class="at">--version</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="ex">fasttree</span> <span class="at">-help</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="ex">rpsblast</span> <span class="at">-version</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
</div>
</div>
</div>
<section id="preparing-your-reference-database" class="level2">
<h2 class="anchored" data-anchor-id="preparing-your-reference-database">Preparing Your Reference Database</h2>
<p>What genomes are you searching? Before searching for genes, you need to prepare a combined reference database from your genome files. If you are searching just a single genome, you can proceed to the next section.</p>
<section id="combine-genome-protein-files" class="level3">
<h3 class="anchored" data-anchor-id="combine-genome-protein-files">Combine Genome Protein Files</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine all .faa files from your genome database into a single file</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> input/genome_database/<span class="pp">*</span>/<span class="pp">*</span>.faa <span class="op">&gt;</span> genome_database/concatenated_orfs.faa</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Verify the combined file</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-c</span> <span class="st">"^&gt;"</span> genome_database/concatenated_orfs.faa  <span class="co"># Count number of sequences</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="alternative-combine-with-custom-glob-pattern" class="level3">
<h3 class="anchored" data-anchor-id="alternative-combine-with-custom-glob-pattern">Alternative: Combine with Custom Glob Pattern</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># If your files are in a different location or pattern</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> path/to/genomes/<span class="pp">**</span>/<span class="pp">*</span>.faa <span class="op">&gt;</span> concatenated_orfs.faa</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Or combine specific genome files</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> genome1.faa genome2.faa genome3.faa <span class="op">&gt;</span> concatenated_orfs.faa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>Note</strong>: Ensure sequence headers are unique across all input files to avoid ambiguity in search results.</p>
</section>
</section>
<section id="getting-your-query-gene" class="level2">
<h2 class="anchored" data-anchor-id="getting-your-query-gene">Getting Your Query Gene</h2>
<section id="step-1-find-your-taxon-and-annotation" class="level3">
<h3 class="anchored" data-anchor-id="step-1-find-your-taxon-and-annotation">Step 1: Find Your Taxon and Annotation</h3>
<p>Go to one of these resources to find your taxon of interest: - https://www.ebi.ac.uk/interpro/taxonomy/uniprot/<br>
- https://www.kegg.jp/kegg/tables/br08606.html</p>
<p>Then search within that taxon for your annotation of interest.</p>
</section>
<section id="step-2-decide-on-search-strategy" class="level3">
<h3 class="anchored" data-anchor-id="step-2-decide-on-search-strategy">Step 2: Decide on Search Strategy</h3>
<p><strong>Question:</strong> Am I searching broadly for this protein coding gene or specifically?</p>
<ul>
<li><strong>Specifically</strong>: Search the exact sequence - best for closely related genomes with high sequence similarity</li>
<li><strong>By orthologous group</strong>: Use pre-built PSSM profiles from NCBI’s COG database - best when searching for functional orthologs across different species</li>
<li><strong>Broadly</strong>: Get the UniRef50 cluster for that gene and generate an alignment and HMM of all sequences in the cluster - best for capturing diverse homologs while maintaining some specificity</li>
</ul>
</section>
</section>
<section id="specific-search-approach-sequence-based" class="level2">
<h2 class="anchored" data-anchor-id="specific-search-approach-sequence-based">Specific Search Approach (Sequence-based)</h2>
<section id="get-the-uniref90-representative-sequence" class="level3">
<h3 class="anchored" data-anchor-id="get-the-uniref90-representative-sequence">1. Get the UniRef90 Representative Sequence</h3>
<p>You can search for the exact sequence you are interested in. Or, it could be more reproducible to search the UniRef90 representative sequence of your gene of interest:</p>
<p>If going the Uniref90 route, you can search for your sequence at https://www.uniprot.org/blast and get the uniref90 cluster for that sequence.</p>
<p>Or if you got your sequence from kegg.jp, you can click on the uniprot cross reference, then on “similar proteins” then on the Uniref_90 cluster.</p>
<p>Then, download the uniref_90 representative sequence.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace UniRef90_D5H0P4 with your cluster ID from Uniprot</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="st">"https://rest.uniprot.org/uniref/UniRef90_D5H0P4.fasta"</span> <span class="at">-o</span> UniRef90_D5H0P4.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="search-against-genome-database-using-mmseqs2" class="level3">
<h3 class="anchored" data-anchor-id="search-against-genome-database-using-mmseqs2">2. Search Against Genome Database Using MMseqs2</h3>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mmseqs</span> easy-search <span class="dt">\</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  input/UniRef90_D5H0P4.faa <span class="dt">\</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  input/genome_database/concatenated_orfs.faa <span class="dt">\</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  UniRef90_D5H0P4_hits.tsv <span class="dt">\</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">-e</span> 1e-20 <span class="dt">\</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  tmp <span class="dt">\</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">--format-output</span> <span class="st">"query,target,pident,alnlen,evalue"</span> <span class="dt">\</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--format-mode</span> 4</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="domain-based-search-approach-using-ncbi-cog-pssms" class="level2">
<h2 class="anchored" data-anchor-id="domain-based-search-approach-using-ncbi-cog-pssms">Domain-based Search Approach (Using NCBI COG PSSMs)</h2>
<p>This approach uses pre-built Position-Specific Scoring Matrices (PSSMs) from NCBI’s COG (Clusters of Orthologous Genes) database.</p>
<p><strong>Note: COGs can be <em>very</em> broad. Make sure the COG you are looking for is specific enough to the function you are interested in. For reference, there are fewer than 5,000 COGs in the 2024 database.</strong></p>
<section id="download-the-cog-database" class="level3">
<h3 class="anchored" data-anchor-id="download-the-cog-database">1. Download the COG Database</h3>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download the COG database</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="ex">update_blastdb.pl</span> <span class="at">--decompress</span> Cog</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="identify-your-cog-of-interest" class="level3">
<h3 class="anchored" data-anchor-id="identify-your-cog-of-interest">2. Identify Your COG of Interest</h3>
<p>Find the Uniprot protein ID and search for it on http://eggnog5.embl.de</p>
<p>Note the COG ID (e.g., COG0001, COG4716) for reference.</p>
</section>
<section id="search-against-genome-database-using-rps-blast" class="level3">
<h3 class="anchored" data-anchor-id="search-against-genome-database-using-rps-blast">3. Search Against Genome Database Using RPS-BLAST</h3>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Search proteins against the COG database</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="ex">rpsblast</span> <span class="dt">\</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">-query</span> genome_database/concatenated_orfs.faa <span class="dt">\</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">-db</span> Cog <span class="dt">\</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">-out</span> COG_hits.txt <span class="dt">\</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">-evalue</span> 1e-5 <span class="dt">\</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">-outfmt</span> <span class="st">"6 qseqid sseqid pident length evalue bitscore stitle"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="filter-for-your-specific-cog" class="level3">
<h3 class="anchored" data-anchor-id="filter-for-your-specific-cog">4. Filter for Your Specific COG</h3>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract hits for a specific COG (e.g., COG0001)</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="st">"COG0001"</span> COG_hits.txt <span class="op">&gt;</span> COG0001_hits.txt</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract sequence IDs</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f1</span> COG0001_hits.txt <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-u</span> <span class="op">&gt;</span> cog_hit_ids.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="broad-search-approach-hmm-based" class="level2">
<h2 class="anchored" data-anchor-id="broad-search-approach-hmm-based">Broad Search Approach (HMM-based)</h2>
<section id="get-query-sequences-from-uniref50-cluster" class="level3">
<h3 class="anchored" data-anchor-id="get-query-sequences-from-uniref50-cluster">1. Get Query Sequences from UniRef50 Cluster</h3>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace UniRef50_C2EQ21 with your cluster ID</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> <span class="st">"https://rest.uniprot.org/uniprotkb/stream?format=fasta&amp;query=uniref_cluster_50:UniRef50_C2EQ21"</span> <span class="dt">\</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="op">&gt;</span> UniRef50_C2EQ21_all_members.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="align-the-sequences" class="level3">
<h3 class="anchored" data-anchor-id="align-the-sequences">2. Align the Sequences</h3>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mafft</span> UniRef50_C2EQ21_all_members.fasta <span class="op">&gt;</span> UniRef50_C2EQ21_aligned.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="build-the-hmm-profile" class="level3">
<h3 class="anchored" data-anchor-id="build-the-hmm-profile">3. Build the HMM Profile</h3>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">hmmbuild</span> UniRef50_C2EQ21.hmm UniRef50_C2EQ21_aligned.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="search-against-genome-database-using-hmmsearch" class="level3">
<h3 class="anchored" data-anchor-id="search-against-genome-database-using-hmmsearch">4. Search Against Genome Database Using hmmsearch</h3>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">hmmsearch</span> <span class="dt">\</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--tblout</span> UniRef50_C2EQ21_hits.tbl <span class="dt">\</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">-E</span> 1e-10 <span class="dt">\</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  UniRef50_C2EQ21.hmm <span class="dt">\</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  genome_database/concatenated_orfs.faa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="example-downstream-analysis" class="level2">
<h2 class="anchored" data-anchor-id="example-downstream-analysis">Example Downstream Analysis</h2>
<p>Once you have your table of gene hits, you might do a lot of different things with them.</p>
<p>One example would be to align the gene hits together and make a tree.</p>
<p>To do this we’ll parse the gene hit table to get the hit ids, then use seqkit to get the sequences from the reference database we searched. After we have the sequences we’ll align with mafft, and then make a tree with fasttree which can be viewed on ITOL.</p>
<section id="extract-hit-ids" class="level3">
<h3 class="anchored" data-anchor-id="extract-hit-ids">1. Extract Hit IDs</h3>
<p>For MMseqs2 results:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f2</span> UniRef90_D5H0P4_hits.tsv <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-u</span> <span class="op">&gt;</span> hit_ids.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>For RPS-BLAST results:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Already extracted above as cog_hit_ids.txt</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Or extract from the full results:</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cut</span> <span class="at">-f1</span> COG_hits.txt <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-u</span> <span class="op">&gt;</span> hit_ids.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>For HMM search results:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Skip header lines and extract target names</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">grep</span> <span class="at">-v</span> <span class="st">'^#'</span> UniRef50_C2EQ21_hits.tbl <span class="kw">|</span> <span class="fu">awk</span> <span class="st">'{print $1}'</span> <span class="kw">|</span> <span class="fu">sort</span> <span class="at">-u</span> <span class="op">&gt;</span> hit_ids.txt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="retrieve-hit-sequences" class="level3">
<h3 class="anchored" data-anchor-id="retrieve-hit-sequences">2. Retrieve Hit Sequences</h3>
<p>To add the filename (strain ID) to each sequence header:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> file <span class="kw">in</span> genome_database/<span class="pp">*</span>/<span class="pp">*</span>.faa<span class="kw">;</span> <span class="cf">do</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="ex">seqkit</span> grep <span class="at">-f</span> hit_ids.txt <span class="st">"</span><span class="va">$file</span><span class="st">"</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>    <span class="ex">seqkit</span> replace <span class="at">-p</span> <span class="st">'^'</span> <span class="at">-r</span> <span class="st">"</span><span class="va">$(</span><span class="fu">basename</span> <span class="va">${file</span><span class="op">%</span>.faa<span class="va">})</span><span class="st">_"</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="cf">done</span> <span class="op">&gt;</span> hit_sequences.faa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Or without filename prefix:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="ex">seqkit</span> grep <span class="at">-f</span> hit_ids.txt genome_database/<span class="pp">*</span>/<span class="pp">*</span>.faa <span class="op">&gt;</span> hit_sequences.faa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="align-hit-sequencess" class="level3">
<h3 class="anchored" data-anchor-id="align-hit-sequencess">3. Align Hit Sequencess</h3>
<p>Optional: Consider adding reference sequences from UniRef90 before alignment</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="ex">mafft</span> hit_sequences.faa <span class="op">&gt;</span> aligned_hit_sequences.faa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="build-phylogenetic-tree" class="level3">
<h3 class="anchored" data-anchor-id="build-phylogenetic-tree">4. Build Phylogenetic Tree</h3>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="ex">fasttree</span> aligned_hit_sequences.faa <span class="op">&gt;</span> hit_sequences.tree</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
</section>
<section id="notes" class="level2">
<h2 class="anchored" data-anchor-id="notes">Notes</h2>
<section id="e-value-thresholds" class="level3">
<h3 class="anchored" data-anchor-id="e-value-thresholds">E-value Thresholds</h3>
<ul>
<li><strong>MMseqs2</strong>: <code>-e 1e-20</code> is stringent for specific searches</li>
<li><strong>hmmsearch</strong>: <code>-E 1e-10</code> is appropriate for HMM profiles</li>
<li><strong>RPS-BLAST</strong>: <code>-evalue 1e-5</code> is standard for domain searches</li>
<li>Adjust thresholds based on your specific use case and desired sensitivity/specificity</li>
</ul>
</section>
<section id="performance-tips" class="level3">
<h3 class="anchored" data-anchor-id="performance-tips">Performance Tips</h3>
<ul>
<li>The <code>tmp</code> directory in MMseqs2 commands is used for temporary files and will be created automatically</li>
<li>For large datasets, use MAFFT’s <code>--thread</code> option to specify multiple CPU cores</li>
<li>The COG database requires ~500MB of disk space</li>
</ul>
</section>
<section id="cog-database-updates" class="level3">
<h3 class="anchored" data-anchor-id="cog-database-updates">COG Database Updates</h3>
<ul>
<li>The COG database was updated in 2024, increasing coverage from 1,309 to 2,296 species</li>
<li>Use <code>update_blastdb.pl --showall</code> to see all available databases</li>
</ul>
</section>
<section id="phylogenetic-analysis" class="level3">
<h3 class="anchored" data-anchor-id="phylogenetic-analysis">Phylogenetic Analysis</h3>
<ul>
<li>FastTree produces an approximately-maximum-likelihood tree quickly</li>
<li>For more accurate phylogenies, consider using RAxML or IQ-TREE</li>
</ul>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/kwondry\.github\.io\/documentation\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../../materials/metagenomics/virgo2/index.html" class="pagination-link" aria-label="VIRGO2 Pipeline">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">VIRGO2 Pipeline</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>© 2023. Huge thanks to <a href="https://github.com/andrewpbray">Andrew P. Bray</a> and <a href="https://github.com/clauswilke">Claus O. Wilke</a> for open source code and sharing content.</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>This page is built with <i class="fa-solid fa-heart" aria-label="heart"></i> and <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>




</body></html>