<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Kwon Lab Pipelines - R workshop</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../materials/rnaseq/background/index.html" rel="next">
<link href="../../materials/getting-started/local_R/index.html" rel="prev">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/all.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-0.1.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Kwon Lab Pipelines</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://kwonlab.mgh.harvard.edu/"> 
<span class="menu-text">About the Kwon Lab</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/kwondry/documentation"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <a class="flex-grow-1 no-decor" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
          <h1 class="quarto-secondary-nav-title">R workshop</h1>
        </a>     
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Getting Started</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../materials/getting-started/o2/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"> Get your O2 account</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../materials/getting-started/installation/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"> Installing software for work on O2</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">R and R Studio</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../materials/getting-started/local_R/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"> RStudio and recommended packages</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../materials/r-tutorial/index.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"> R workshop</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">RNAseq</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../materials/rnaseq/background/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Background</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../materials/rnaseq/walkthrough/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Walkthrough</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true">
 <span class="menu-text">16S</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../materials/16S/background/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Background</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../materials/16S/kwon_sequencing_data_transfer/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">SMB Data Transfer and Demultiplexing</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../materials/16S/16s_pipeline/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">16S Pipeline</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../materials/16S/valencia/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Assigning CSTs with Valencia</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true">
 <span class="menu-text">Metagenomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../materials/metagenomics/background/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Background</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../materials/metagenomics/virgo2/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">VIRGO2 Pipeline</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true">
 <span class="menu-text">Comparative Genomics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../materials/comparative-genomics/gene-search/index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Gene Search</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#slides-intro-to-rstudio-quarto-and-r" id="toc-slides-intro-to-rstudio-quarto-and-r" class="nav-link active" data-scroll-target="#slides-intro-to-rstudio-quarto-and-r">Slides: Intro to RStudio, Quarto, and R</a></li>
  <li><a href="#download-the-instructional-dataset" id="toc-download-the-instructional-dataset" class="nav-link" data-scroll-target="#download-the-instructional-dataset">Download the instructional dataset:</a></li>
  <li><a href="#exercise-1-data-wrangling-and-simple-manipulations" id="toc-exercise-1-data-wrangling-and-simple-manipulations" class="nav-link" data-scroll-target="#exercise-1-data-wrangling-and-simple-manipulations">Exercise 1: Data wrangling and simple manipulations</a>
  <ul class="collapse">
  <li><a href="#the-pipe-read-and-then" id="toc-the-pipe-read-and-then" class="nav-link" data-scroll-target="#the-pipe-read-and-then">The pipe (<code>%&gt;%</code>, read: “and then”)</a></li>
  <li><a href="#choosing-data-rows" id="toc-choosing-data-rows" class="nav-link" data-scroll-target="#choosing-data-rows">Choosing data rows</a></li>
  <li><a href="#filtering-for-multiple-conditions" id="toc-filtering-for-multiple-conditions" class="nav-link" data-scroll-target="#filtering-for-multiple-conditions">Filtering for multiple conditions</a></li>
  <li><a href="#choosing-data-columns" id="toc-choosing-data-columns" class="nav-link" data-scroll-target="#choosing-data-columns">Choosing data columns</a></li>
  <li><a href="#choosing-columns-for-removal" id="toc-choosing-columns-for-removal" class="nav-link" data-scroll-target="#choosing-columns-for-removal">Choosing columns for removal</a></li>
  <li><a href="#sorting-data" id="toc-sorting-data" class="nav-link" data-scroll-target="#sorting-data">Sorting data</a></li>
  <li><a href="#counting" id="toc-counting" class="nav-link" data-scroll-target="#counting">Counting</a></li>
  <li><a href="#chaining-analysis-steps-into-pipelines" id="toc-chaining-analysis-steps-into-pipelines" class="nav-link" data-scroll-target="#chaining-analysis-steps-into-pipelines">Chaining analysis steps into pipelines</a></li>
  </ul></li>
  <li><a href="#exercise-2-data-visualization-activity" id="toc-exercise-2-data-visualization-activity" class="nav-link" data-scroll-target="#exercise-2-data-visualization-activity">Exercise 2: Data Visualization Activity</a>
  <ul class="collapse">
  <li><a href="#genetic-profiles-of-cancer" id="toc-genetic-profiles-of-cancer" class="nav-link" data-scroll-target="#genetic-profiles-of-cancer">Genetic profiles of cancer</a></li>
  </ul></li>
  <li><a href="#exercise-3-ggplot-exercise" id="toc-exercise-3-ggplot-exercise" class="nav-link" data-scroll-target="#exercise-3-ggplot-exercise">Exercise 3: ggplot exercise</a>
  <ul class="collapse">
  <li><a href="#basic-use-of-ggplot" id="toc-basic-use-of-ggplot" class="nav-link" data-scroll-target="#basic-use-of-ggplot">Basic use of ggplot</a></li>
  <li><a href="#more-complex-geoms" id="toc-more-complex-geoms" class="nav-link" data-scroll-target="#more-complex-geoms">More complex geoms</a></li>
  <li><a href="#adding-color" id="toc-adding-color" class="nav-link" data-scroll-target="#adding-color">Adding color</a></li>
  <li><a href="#using-aesthetics-as-parameters" id="toc-using-aesthetics-as-parameters" class="nav-link" data-scroll-target="#using-aesthetics-as-parameters">Using aesthetics as parameters</a></li>
  </ul></li>
  <li><a href="#exercise-4-data-wrangling-2" id="toc-exercise-4-data-wrangling-2" class="nav-link" data-scroll-target="#exercise-4-data-wrangling-2">Exercise 4: Data wrangling 2</a>
  <ul class="collapse">
  <li><a href="#analyzing-subsets" id="toc-analyzing-subsets" class="nav-link" data-scroll-target="#analyzing-subsets">Analyzing subsets</a></li>
  <li><a href="#grouping" id="toc-grouping" class="nav-link" data-scroll-target="#grouping">Grouping</a></li>
  <li><a href="#performing-summaries" id="toc-performing-summaries" class="nav-link" data-scroll-target="#performing-summaries">Performing summaries</a></li>
  <li><a href="#making-tables-wider-or-longer" id="toc-making-tables-wider-or-longer" class="nav-link" data-scroll-target="#making-tables-wider-or-longer">Making tables wider or longer</a></li>
  <li><a href="#combining-datasets-with-joins" id="toc-combining-datasets-with-joins" class="nav-link" data-scroll-target="#combining-datasets-with-joins">Combining datasets with joins</a></li>
  </ul></li>
  <li><a href="#advanced-next-steps" id="toc-advanced-next-steps" class="nav-link" data-scroll-target="#advanced-next-steps">Advanced next steps</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block">R workshop</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<blockquote class="blockquote">
<p>Learn how to use RStudio and the <code>tidyverse</code> for data exploration</p>
</blockquote>
<section id="slides-intro-to-rstudio-quarto-and-r" class="level2">
<h2 class="anchored" data-anchor-id="slides-intro-to-rstudio-quarto-and-r">Slides: Intro to RStudio, Quarto, and R</h2>
<p><a href="../../materials/r-tutorial/slides.html">Make slides full screen</a></p>
<iframe class="slide-deck" src="slides.html" height="420" width="747" style="border: 1px solid #2e3846;"></iframe>
</section>
<section id="download-the-instructional-dataset" class="level2">
<h2 class="anchored" data-anchor-id="download-the-instructional-dataset">Download the instructional dataset:</h2>
<p>The main data set for use in lectures is split into 5 tables. The idea of this dataset is that there is a randomized controlled trial of a drug aimed at reducing HIV risk by reducing inflammation. There were 23 participants in the placebo arm and 21 in the treatment arm. There were 3 visits for the trial – baseline (before any treatment occurred), week_1 (1 week after treatment) and week_7 (7 weeks after treatment). At each time point inflammation was measured using luminex (elisa_cytokines table) and immune cells counts were measured from a pap-smear (flow_cytometry table).</p>
<ul>
<li><a href="datasets/instructional_dataset/00_sample_ids.csv"><code>00_sample_ids.csv</code></a></li>
</ul>
<p><code>pid</code> – participant id<br>
<code>time_point</code> – “baseline”, “week_1”, or “week_7” <code>arm</code> – “treatment” or “placebo”<br>
<code>sample_id</code> – the “wet-lab” sample id associated with this timepoint</p>
<ul>
<li><a href="datasets/instructional_dataset/01_participant_metadata.csv"><code>01_participant_metadata.csv</code></a></li>
</ul>
<p><code>pid</code> - participant id<br>
<code>arm</code> - “treatment” or “placebo”<br>
<code>smoker</code> - “yes” or “no”<br>
<code>age</code> – age in years<br>
<code>education</code> – 4 options (“less than grade 9”, “grade 10-12, not matriculated”, “grade 10-12, matriculated”, “post-secondary”)<br>
<code>sex</code> – all participants are “Female”</p>
<ul>
<li><a href="datasets/instructional_dataset/02_visit_clinical_measurements.csv"><code>02_visit_clinical_measurements.csv</code></a></li>
</ul>
<p><code>pid</code> – particpant id<br>
<code>time_point</code> – “baseline”, “week_1”, or “week_7”<br>
<code>arm</code> – “treatment” or “placebo”<br>
<code>nugent_score</code> – <a href="https://en.wikipedia.org/wiki/Nugent_score">Nugent Score</a>, a number from 0-10. 0-3 is no BV, 4-6 is intermediate BV, and 7-10 is BV . <code>crp_blood</code> – decimal number representing C-reactive protein blood test (CRP) <code>ph</code> – vaginal pH</p>
<ul>
<li><a href="datasets/instructional_dataset/03_elisa_cytokines.csv"><code>03_elisa_cytokines.csv</code></a></li>
</ul>
<p><code>sample_id</code> - the “wet-lab” sample id associated with this timepoint<br>
<code>cytokine</code> - “IL-1a”, “IL-10”, “IL-1b”, “IL-8”, “IL-6”, “TNFa”, “IP-10”, “MIG”, “IFN-Y”, “MIP-3a”<br>
<code>conc</code> – decimal number representing concentration<br>
<code>limits</code> – either “within limits” or “out of range”</p>
<ul>
<li><a href="datasets/instructional_dataset/04_flow_cytometry.csv"><code>04_flow_cytometry.csv</code></a></li>
</ul>
<p><code>sample_id</code> - the “wet-lab” sample id associated with this timepoint<br>
All other columns – the integer count of this type of cell in this sample</p>
<p><code>cd4_t_cells</code> might best be analyzed as a proportion of <code>cd3_t_cells</code>… ### Read in data</p>
<p>Data is often in tables, and the easiest way to store tabular data is in <code>csv</code> or <code>tsv</code> format.</p>
<p><code>csv</code> - comma separated values<br>
<code>tsv</code> - tab separated values</p>
<p>to read in data stored this way use <code>read_csv(filename)</code> or <code>read_tsv(filename)</code></p>
</section>
<section id="exercise-1-data-wrangling-and-simple-manipulations" class="level2">
<h2 class="anchored" data-anchor-id="exercise-1-data-wrangling-and-simple-manipulations">Exercise 1: Data wrangling and simple manipulations</h2>
<p>In this worksheet, we will discuss how to perform basic data manipulations, such as filtering data rows that meet certain conditions, choosing data columns, and arranging data in ascending or descending order.</p>
<p>We will be using the R package, <strong>tidyverse</strong> for the data manipulation functions <code>%&gt;%</code>, <code>filter()</code>, <code>select()</code>, <code>arrange()</code>, <code>count()</code>, and <code>mutate()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="in">#| eval: false</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="in">#| warning: false</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="in">library(tidyverse)</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="in">table_01 &lt;- read_csv("data/01_participant_metadata.csv")</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="in">table_02 &lt;- read_csv("data/02_visit_clinical_measurements.csv")</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="the-pipe-read-and-then" class="level3">
<h3 class="anchored" data-anchor-id="the-pipe-read-and-then">The pipe (<code>%&gt;%</code>, read: “and then”)</h3>
<p>When writing complex data analysis pipelines, we frequently use the pipe operator <code>%&gt;%</code> to move data from one analysis step to the next. The pipe is pronounced “and then”, and it takes the data on its left and uses it as the first argument for the function on its right.</p>
<p>For example, to see the first few lines of a dataset, we often write <code>head(data)</code>. Instead, we can write <code>data %&gt;% head()</code>.</p>
<p>Try this yourself. Write code that displays the first few lines of <code>table_01</code>, using <code>%&gt;%</code> and <code>head()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>table_01 ___ <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="choosing-data-rows" class="level3">
<h3 class="anchored" data-anchor-id="choosing-data-rows">Choosing data rows</h3>
<p>The function <code>filter()</code> allows you to find rows in a dataset that meet one or more specific conditions. The syntax is <code>data %&gt;% filter(condition)</code>, where condition is a logical condition. For example, <code>filter(x &gt; 5)</code> would pick all rows for which the value in column <code>x</code> is greater than 5.</p>
<p>Below, write code to pick all measurements from the <code>placebo</code> arm in <code>table_02</code> :</p>
<div class="cell" data-exercise="true">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>table_02 <span class="sc">%&gt;%</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(___)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="filtering-for-multiple-conditions" class="level3">
<h3 class="anchored" data-anchor-id="filtering-for-multiple-conditions">Filtering for multiple conditions</h3>
<p>You can also state multiple conditions, separated by a comma. For example, <code>filter(x &gt; 5, y &lt; 2)</code> would pick all rows for which the value in the column <code>x</code> is greater than 5 and the value in the column <code>y</code> is less than 2. Note that the conditions are combined via logical and, both need to be satisfied for the row to be picked.</p>
<p>To try this out, pick all measurements at <code>baseline</code> for the <code>placebo</code> arm in <code>table_02</code> :</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>table_02 <span class="sc">%&gt;%</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(time_point <span class="sc">==</span> ___, arm <span class="sc">==</span> ___)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="choosing-data-columns" class="level3">
<h3 class="anchored" data-anchor-id="choosing-data-columns">Choosing data columns</h3>
<p>The function <code>select()</code> allows you to pick specific data columns by name. This is frequently useful when a dataset has many more columns than we are interested in at the time. For example, if we are only interested in the participants’ education, and sex, we could select these three columns:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>table_01 <span class="sc">%&gt;%</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(education, sex)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 44 × 2
   education                     sex   
   &lt;chr&gt;                         &lt;chr&gt; 
 1 grade 10-12, matriculated     Female
 2 grade 10-12, matriculated     Female
 3 post-secondary                Female
 4 grade 10-12, not matriculated Female
 5 grade 10-12, matriculated     Female
 6 post-secondary                Female
 7 grade 10-12, not matriculated Female
 8 grade 10-12, not matriculated Female
 9 grade 10-12, not matriculated Female
10 less than grade 9             Female
# ℹ 34 more rows</code></pre>
</div>
</div>
<p>Try this yourself, picking the columns representing the participant age (<code>age</code>), and then study arm (<code>arm</code>).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>table_01 <span class="sc">%&gt;%</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(___)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="choosing-columns-for-removal" class="level3">
<h3 class="anchored" data-anchor-id="choosing-columns-for-removal">Choosing columns for removal</h3>
<p>Another situation that arises frequently is one where we want to remove specific columns. We can also do this with <code>select()</code>, but now write <code>select(-column)</code> to remove one or more columns.</p>
<p>Try this. Remove the column <code>smoker</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>table_01 <span class="sc">%&gt;%</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(___)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And now remove both <code>smoker</code> and <code>education</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>table_01 <span class="sc">%&gt;%</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>___, <span class="sc">-</span>___)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sorting-data" class="level3">
<h3 class="anchored" data-anchor-id="sorting-data">Sorting data</h3>
<p>The function <code>arrange()</code> allows you to sort data by one or more columns. For example, <code>data %&gt;% arrange(x)</code> would sort the data by increasing values of <code>x</code>, and <code>data %&gt;% arrange(x, y)</code> would sort the data first by <code>x</code> and then, for ties in <code>x</code>, by <code>y</code>.</p>
<p>Sort the participants by age:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>table_01 <span class="sc">%&gt;%</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(___)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To arrange data in descending order, enclose the data column in <code>desc()</code>. For example, <code>data %&gt;% arrange(desc(x))</code> would sort the data by decreasing values of <code>x</code>. (<code>desc</code> stands for “descending”.)</p>
</section>
<section id="counting" class="level3">
<h3 class="anchored" data-anchor-id="counting">Counting</h3>
<p>We frequently want to count how many times a particular value or combination of values occurs in a dataset. We do this using the <code>count()</code> function. For example, the following code counts how many participants there are in each study arm in table_01.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>table_01 <span class="sc">%&gt;%</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(arm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  arm           n
  &lt;chr&gt;     &lt;int&gt;
1 placebo      23
2 treatment    21</code></pre>
</div>
</div>
<p>Now try this yourself. Count how many smokers and non smokers there are.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>table_01 <span class="sc">%&gt;%</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(___)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can also perform more fine-grained counts, by providing the <code>count()</code> function with more than one column name. See if you can count how many smokers and non smokers the dataset contains for each treatment arm</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>table_01 <span class="sc">%&gt;%</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(___, smoker)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now count how many participants are in each study arm for each education level in the dataset</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>table_01 <span class="sc">%&gt;%</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(___, arm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="chaining-analysis-steps-into-pipelines" class="level3">
<h3 class="anchored" data-anchor-id="chaining-analysis-steps-into-pipelines">Chaining analysis steps into pipelines</h3>
<p>We can chain multiple analysis steps into a pipeline by continuing to add “and then” statements. For example, <code>data %&gt;% count(...) %&gt;% arrange(...)</code> would first count and then sort the data.</p>
<p>Try this out by counting the number of participants in each study arm and then sorting by the number of participants</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>table_01 <span class="sc">%&gt;%</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">count</span>(___) <span class="sc">%&gt;%</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(___)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br>
</p>
</section>
</section>
<section id="exercise-2-data-visualization-activity" class="level2">
<h2 class="anchored" data-anchor-id="exercise-2-data-visualization-activity">Exercise 2: Data Visualization Activity</h2>
<p><em>reproduced from a Schmidt Science Fellowship activity, thanks to Fatima Hussain</em></p>
<blockquote class="blockquote">
<p>Patterns are the essence of data exploration and our eyes’ ability to pick them out is integral to data understanding. Much of the data we work with, however, do not have a natural form and we need to make decisions about how they are to be represented. Try different ways to visualize the datasets so meaningful patterns may be found.</p>
</blockquote>
<section id="genetic-profiles-of-cancer" class="level3">
<h3 class="anchored" data-anchor-id="genetic-profiles-of-cancer">Genetic profiles of cancer</h3>
<p>These datasets contains 10 cancer samples. <a href="#tbl-1" class="quarto-xref">Table&nbsp;1</a> describes the mutational status for a set of genes (A-E) and whether a mutation if absent (0) or present (1). <a href="#tbl-2" class="quarto-xref">Table&nbsp;2</a> summarizes the expression levels of those genes, ranging from no expression (0) to high expression (3).</p>
<div id="tbl-1" class="striped table-bordered table-sm quarto-float anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;1: Mutational status for a set of genes
</figcaption>
<div aria-describedby="tbl-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="table-striped table-bordered table-sm small table">
<colgroup>
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 9%">
</colgroup>
<thead>
<tr class="header">
<th>sample -&gt;</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
<th>7</th>
<th>8</th>
<th>9</th>
<th>10</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Gene A</strong></td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td><strong>Gene B</strong></td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>0</td>
<td>1</td>
<td>1</td>
</tr>
<tr class="odd">
<td><strong>Gene C</strong></td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>1</td>
<td>1</td>
<td>1</td>
</tr>
<tr class="even">
<td><strong>Gene D</strong></td>
<td>1</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="odd">
<td><strong>Gene E</strong></td>
<td>0</td>
<td>1</td>
<td>1</td>
<td>0</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>1</td>
<td>0</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
<div id="tbl-2" class="striped table-bordered table-sm quarto-float anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;2: Expression levels for a set of genes
</figcaption>
<div aria-describedby="tbl-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<table class="table-striped table-bordered table-sm small table">
<colgroup>
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 9%">
<col style="width: 9%">
</colgroup>
<thead>
<tr class="header">
<th>sample -&gt;</th>
<th>1</th>
<th>2</th>
<th>3</th>
<th>4</th>
<th>5</th>
<th>6</th>
<th>7</th>
<th>8</th>
<th>9</th>
<th>10</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Gene A</strong></td>
<td>2</td>
<td>1</td>
<td>1</td>
<td>2</td>
<td>2</td>
<td>0</td>
<td>2</td>
<td>1</td>
<td>1</td>
<td>2</td>
</tr>
<tr class="even">
<td><strong>Gene B</strong></td>
<td>1</td>
<td>1</td>
<td>2</td>
<td>1</td>
<td>0</td>
<td>0</td>
<td>0</td>
<td>2</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="odd">
<td><strong>Gene C</strong></td>
<td>1</td>
<td>1</td>
<td>3</td>
<td>1</td>
<td>2</td>
<td>2</td>
<td>3</td>
<td>0</td>
<td>3</td>
<td>0</td>
</tr>
<tr class="even">
<td><strong>Gene D</strong></td>
<td>0</td>
<td>0</td>
<td>2</td>
<td>1</td>
<td>3</td>
<td>3</td>
<td>2</td>
<td>1</td>
<td>1</td>
<td>1</td>
</tr>
<tr class="odd">
<td><strong>Gene E</strong></td>
<td>1</td>
<td>3</td>
<td>3</td>
<td>1</td>
<td>3</td>
<td>1</td>
<td>2</td>
<td>1</td>
<td>3</td>
<td>2</td>
</tr>
</tbody>
</table>
</div>
</figure>
</div>
<p><br>
</p>
<div class="line-block">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;1. Think about the problem on your own for 5 minutes.<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;2. In your groups, discuss and create different visualizations to highlight underlying patterns<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3. Summarize the group’s approach<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;4. Elect/volunteer a spokesperson to present the solution</div>
<p><br>
</p>
<p>On <a href="excalidraw.com">excalidraw.com</a> or pen and paper, create a visualization to answer the questions below. Check out <a href="https://rstudio.github.io/cheatsheets/data-visualization.pdf">the data visualization cheatsheet</a> for inspiration. Think about values you’d have to compute, or if you have everything you need already in our <code>joined_data</code> table.</p>
<p><strong>Consider the following concepts when creating your visualizations</strong></p>
</section>
<div class="grid">
<div class="g-col-12 text-center">
<section id="patterns" class="level3">
<h3 class="anchored" data-anchor-id="patterns">Patterns</h3>
<p>Patterns are the essence of data exploration. What kinds of representation will produce the most meaningful insights?</p>
<p><img src="images/ws_patterns1.png" class="img-fluid" width="200"> <img src="images/ws_patterns2.png" class="img-fluid" width="200"> <img src="images/ws_patterns3.png" class="img-fluid" width="200"> <img src="images/ws_patterns4.png" class="img-fluid" width="200"></p>
</section>
<section id="encodings" class="level3">
<h3 class="anchored" data-anchor-id="encodings">Encodings</h3>
<p>Some visual estimations are easier to make than others. How might you use <strong>encodings</strong> that are less accurate but otherwise better at conveying overall trends?</p>
<p><img src="images/ws_encod1.png" class="img-fluid" style="width:45.0%"> <img src="images/ws_encod2.png" class="img-fluid" style="width:45.0%"> <img src="images/ws_encod3.png" class="img-fluid" style="width:45.0%"></p>
</section>
</div>
<div class="g-col-12 text-center">
<section id="color" class="level3">
<h3 class="anchored" data-anchor-id="color">Color</h3>
<p><strong>Color</strong> is a powerful encoding that presents several challenges. Have you chosen a color scale that is optimal for that data type?</p>
<p><img src="images/ws_color1.png" class="img-fluid" style="width:19.0%"> <img src="images/ws_color2.png" class="img-fluid" style="width:19.0%"> <img src="images/ws_color3.png" class="img-fluid" style="width:19.0%"> <img src="images/ws_color4.png" class="img-fluid" style="width:19.0%"></p>
</section>
<section id="salience-and-relevance" class="level3">
<h3 class="anchored" data-anchor-id="salience-and-relevance">Salience and Relevance</h3>
<p>Pop-out effects enable quick recognition. Are the most <strong>noticeable</strong> elements of your visualizations also the most <strong>relevant</strong>?</p>
<p><img src="images/ws_salience2.png" class="img-fluid" style="width:90.0%"></p>
</section>
</div>
</div>
</section>
<section id="exercise-3-ggplot-exercise" class="level2">
<h2 class="anchored" data-anchor-id="exercise-3-ggplot-exercise">Exercise 3: ggplot exercise</h2>
<p>For this exercise we’ll be using the instructional dataset. <a href="../../../datasets.html">Download the dataset here</a>.</p>
<p>In this worksheet, we will discuss a core concept of ggplot, the mapping of data values onto aesthetics.</p>
<p>We will be using the R package <code>tidyverse</code>, which includes <code>ggplot()</code> and related functions.</p>
<p>Copy the following code chunk into your quarto document. Notice the error in the <code>read_csv()</code> line - it wants you to supply the <code>file</code> name to read. Fix the error!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse) <span class="co"># load the tidyverse library</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co"># we want to use the data in the visit_clinical_measurements file</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>clinical_measurements <span class="ot">&lt;-</span> <span class="fu">read_csv</span>() <span class="co"># read in your data 02_visit_clinical_measurements.csv</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error in read_csv(): argument "file" is missing, with no default</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co">#then show the first few rows</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(clinical_measurements)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-error">
<pre><code>Error: object 'clinical_measurements' not found</code></pre>
</div>
</div>
<section id="basic-use-of-ggplot" class="level3">
<h3 class="anchored" data-anchor-id="basic-use-of-ggplot">Basic use of ggplot</h3>
<p>In the most basic use of ggplot, we call the <code>ggplot()</code> function with a dataset and an aesthetic mapping (created with <code>aes()</code>), and then we add a geom, such as <code>geom_line()</code> to draw lines or <code>geom_point()</code> to draw points.</p>
<p>Try this for yourself. Map the column <code>ph</code> onto the x axis and the column <code>crp_blood</code> onto the y axis, and use <code>geom_line()</code> to display the data.</p>
<p>Whenever you see <code>___</code> in the code below, that means you should swap it in with your own code.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(clinical_measurements, <span class="fu">aes</span>(<span class="at">x =</span> ___, <span class="at">y =</span> ___)) <span class="sc">+</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">___</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Try again. Now use <code>geom_point()</code> instead of <code>geom_line()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(clinical_measurements, <span class="fu">aes</span>(<span class="at">x =</span> ___, <span class="at">y =</span> ___)) <span class="sc">+</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">___</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And now swap which column you map to x and which to y.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(clinical_measurements, <span class="fu">aes</span>(<span class="at">x =</span> ___, <span class="at">y =</span> ___)) <span class="sc">+</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">___</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="more-complex-geoms" class="level3">
<h3 class="anchored" data-anchor-id="more-complex-geoms">More complex geoms</h3>
<p>You can use other geoms to make different types of plots. For example, <code>geom_boxplot()</code> will make boxplots. For boxplots, we frequently want categorical data on the x or y axis. For example, we might want a separate boxplot for each month. Try this out. Put <code>nugent_score</code> on the x axis, <code>ph</code> on the y axis, and use <code>geom_boxplot()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(clinical_measurements, <span class="fu">aes</span>(<span class="at">x =</span> ___, <span class="at">y =</span> ___)) <span class="sc">+</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">___</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now try swapping the x and y axis <code>geom_jitter()</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(clinical_measurements, <span class="fu">aes</span>(<span class="at">x =</span> ___, <span class="at">y =</span> ___)) <span class="sc">+</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">___</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now try swapping the x and y axis</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(clinical_measurements, <span class="fu">aes</span>(<span class="at">x =</span> ___, <span class="at">y =</span> ___)) <span class="sc">+</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">___</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="adding-color" class="level3">
<h3 class="anchored" data-anchor-id="adding-color">Adding color</h3>
<p>Try again with <code>geom_jitter()</code>, this time using <code>ph</code> as the location along the y axis and <code>arm</code> for the color. Remember to check the <code>ggplot</code> cheat sheet, or type <code>?geom_jitter()</code> in the console to and look at the “Aesthetics” section if you get stuck.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(clinical_measurements, <span class="fu">aes</span>(<span class="at">x =</span> ___, <span class="at">y =</span> ___)) <span class="sc">+</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">___</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>(Hint: Try adding <code>size = 3</code> as a parameter to the <code>geom_jitter()</code> to create larger points.)</p>
</section>
<section id="using-aesthetics-as-parameters" class="level3">
<h3 class="anchored" data-anchor-id="using-aesthetics-as-parameters">Using aesthetics as parameters</h3>
<p>Many of the aesthetics (such as <code>color</code>, <code>fill</code>, and also <code>size</code> to change line size or point thickness) can be used as parameters inside a geom rather than inside an <code>aes()</code> statement. The difference is that when you use an aesthetic as a parameter, you specify a specific value, such as <code>color = "blue"</code>, rather than a mapping, such as <code>aes(color = arm)</code>. Notice the difference: Inside the <code>aes()</code> function, we don’t actually specify the specific color values, ggplot does that for us. We only say that we want the data values of the <code>arm</code> column to correspond to different colors. (We will learn later how to tell ggplot to use specific colors in this mapping.)</p>
<p>Try this with the boxplot example from the previous section. Map <code>arm</code> onto the <code>fill</code> aesthetic but set the color of the lines to <code>"navyblue"</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(clinical_measurements, <span class="fu">aes</span>(<span class="at">x =</span> ___, <span class="at">y =</span> ___)) <span class="sc">+</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">___</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now do the reverse. Map arm onto the line colors of the boxplot, but will the box with the color <code>"navyblue"</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(clinical_measurements, <span class="fu">aes</span>(<span class="at">x =</span> ___, <span class="at">y =</span> ___)) <span class="sc">+</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">___</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Great, that’s all for now! If you are done, put a yellow sticky note on your laptop!</p>
</section>
</section>
<section id="exercise-4-data-wrangling-2" class="level2">
<h2 class="anchored" data-anchor-id="exercise-4-data-wrangling-2">Exercise 4: Data wrangling 2</h2>
<p>In this exercise, we will continue with basic data manipulations, now moving on to grouping and summarizing, making data tables wider or longer, and joining data tables.</p>
<p>We will be using the R package, <strong>tidyverse</strong> for the data manipulation functions <code>%&gt;%</code>, <code>group_by()</code>, <code>summarize()</code>, <code>pivot_wider()</code>, <code>pivot_longer()</code>, and join functions such as <code>left_join()</code></p>
<p>Paste the following into the top code chunk of your <code>qmd</code> file.</p>
<p>Download the data files from the <a href="../../../datasets.qmd">dataset page</a> and place all 5 files into this directory.</p>
<p>Paste the following code chunk into a new qmd file in this project:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>table_01 <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"01_participant_metadata.csv"</span>)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>table_02 <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"02_visit_clinical_measurements.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="analyzing-subsets" class="level3">
<h3 class="anchored" data-anchor-id="analyzing-subsets">Analyzing subsets</h3>
<p>In many data analysis settings, we want to break a dataset into subsets and then perform some summary calculation on each subset. The simplest example is counting, which we have done previously with the count() function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="in">table_01 %&gt;%</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="in">  count(arm)</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 2
  arm           n
  &lt;chr&gt;     &lt;int&gt;
1 placebo      23
2 treatment    21</code></pre>
</div>
</div>
<p>This function subdivides the penguins dataset into subsets for each species and then calculates the number n for each subset, which is the number of observations in each subset.</p>
<p>The function <code>count()</code> here does all the work for us, but what if instead of counting we wanted to calculate the mean weight of the penguins for each species, or calculate the mean weight and count at the same time? We need a general framework that allows us to do these kinds of calculations with maximum flexibility.</p>
<p>The tidyverse approach is to first group a dataset with <code>group_by()</code> and then to calculate grouped summaries with <code>summarize()</code>.</p>
</section>
<section id="grouping" class="level3">
<h3 class="anchored" data-anchor-id="grouping">Grouping</h3>
<p>Let’s first consider just grouping. If we look at the raw R output of just the penguins table or the penguins table after running it through <code>group_by(arm)</code>, we see that the table is the same, except in the second case there is a line <code># Groups: arm [2]</code> which indicates that the table is grouped by arm and there are two groups. (Here, we need to pipe the tables into the print() function to see the raw R output instead of a formatted table that would hide the grouping information.)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>table_01 <span class="sc">%&gt;%</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 44 × 6
   pid    arm       smoker       age education                     sex   
   &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt;                         &lt;chr&gt; 
 1 pid_01 placebo   non-smoker    26 grade 10-12, matriculated     Female
 2 pid_02 placebo   smoker        33 grade 10-12, matriculated     Female
 3 pid_03 placebo   smoker        30 post-secondary                Female
 4 pid_04 placebo   non-smoker    34 grade 10-12, not matriculated Female
 5 pid_05 treatment non-smoker    29 grade 10-12, matriculated     Female
 6 pid_06 placebo   smoker        34 post-secondary                Female
 7 pid_07 placebo   non-smoker    31 grade 10-12, not matriculated Female
 8 pid_08 placebo   smoker        30 grade 10-12, not matriculated Female
 9 pid_09 treatment non-smoker    35 grade 10-12, not matriculated Female
10 pid_10 treatment non-smoker    32 less than grade 9             Female
# ℹ 34 more rows</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>table_01 <span class="sc">%&gt;%</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(arm) <span class="sc">%&gt;%</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 44 × 6
# Groups:   arm [2]
   pid    arm       smoker       age education                     sex   
   &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt;                         &lt;chr&gt; 
 1 pid_01 placebo   non-smoker    26 grade 10-12, matriculated     Female
 2 pid_02 placebo   smoker        33 grade 10-12, matriculated     Female
 3 pid_03 placebo   smoker        30 post-secondary                Female
 4 pid_04 placebo   non-smoker    34 grade 10-12, not matriculated Female
 5 pid_05 treatment non-smoker    29 grade 10-12, matriculated     Female
 6 pid_06 placebo   smoker        34 post-secondary                Female
 7 pid_07 placebo   non-smoker    31 grade 10-12, not matriculated Female
 8 pid_08 placebo   smoker        30 grade 10-12, not matriculated Female
 9 pid_09 treatment non-smoker    35 grade 10-12, not matriculated Female
10 pid_10 treatment non-smoker    32 less than grade 9             Female
# ℹ 34 more rows</code></pre>
</div>
</div>
<p>We can also group by multiple data columns at once, and we can undo any grouping with <code>ungroup()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>table_01 <span class="sc">%&gt;%</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(arm, education) <span class="sc">%&gt;%</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 44 × 6
# Groups:   arm, education [8]
   pid    arm       smoker       age education                     sex   
   &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt;                         &lt;chr&gt; 
 1 pid_01 placebo   non-smoker    26 grade 10-12, matriculated     Female
 2 pid_02 placebo   smoker        33 grade 10-12, matriculated     Female
 3 pid_03 placebo   smoker        30 post-secondary                Female
 4 pid_04 placebo   non-smoker    34 grade 10-12, not matriculated Female
 5 pid_05 treatment non-smoker    29 grade 10-12, matriculated     Female
 6 pid_06 placebo   smoker        34 post-secondary                Female
 7 pid_07 placebo   non-smoker    31 grade 10-12, not matriculated Female
 8 pid_08 placebo   smoker        30 grade 10-12, not matriculated Female
 9 pid_09 treatment non-smoker    35 grade 10-12, not matriculated Female
10 pid_10 treatment non-smoker    32 less than grade 9             Female
# ℹ 34 more rows</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>table_01 <span class="sc">%&gt;%</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(arm, education) <span class="sc">%&gt;%</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 44 × 6
   pid    arm       smoker       age education                     sex   
   &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt;                         &lt;chr&gt; 
 1 pid_01 placebo   non-smoker    26 grade 10-12, matriculated     Female
 2 pid_02 placebo   smoker        33 grade 10-12, matriculated     Female
 3 pid_03 placebo   smoker        30 post-secondary                Female
 4 pid_04 placebo   non-smoker    34 grade 10-12, not matriculated Female
 5 pid_05 treatment non-smoker    29 grade 10-12, matriculated     Female
 6 pid_06 placebo   smoker        34 post-secondary                Female
 7 pid_07 placebo   non-smoker    31 grade 10-12, not matriculated Female
 8 pid_08 placebo   smoker        30 grade 10-12, not matriculated Female
 9 pid_09 treatment non-smoker    35 grade 10-12, not matriculated Female
10 pid_10 treatment non-smoker    32 less than grade 9             Female
# ℹ 34 more rows</code></pre>
</div>
</div>
<p>Now try this yourself. Group the table_01 dataset by education_level and smoker.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>table_01 <span class="sc">%&gt;%</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>  ___ <span class="sc">%&gt;%</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><em>Answer for yourself</em> How many distinct groups did this produce?</p>
<p>Now undo the previous grouping.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="in"># build all the code for this exercise</span></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Also verify what the output looks like when you omit the <code>print()</code> function at the end.</p>
</section>
<section id="performing-summaries" class="level3">
<h3 class="anchored" data-anchor-id="performing-summaries">Performing summaries</h3>
<p>Once we have set up a grouping for a data table, we can then calculate summary data with the <code>summarize()</code> function. This function works similarly to <code>mutate()</code>, in that we provide it with statements of the form <code>&lt;new column name&gt; = &lt;computation&gt;</code>, where <code>&lt;new column name&gt;</code> stands for the name of the new column that is being created and <code>&lt;computation&gt;</code> stands for the computation that is used to generate the values in the new column.</p>
<p>As an example, using <code>table_02</code> we want to calculate the median ph of participants, by <code>arm</code>, we could write <code>summarise(median_ph = median(ph))</code>, and this would create a new column called <code>median_ph</code></p>
<p>Try this out. First group by <code>arm</code> and then make the new column:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>table_02 <span class="sc">%&gt;%</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(____) <span class="sc">%&gt;%</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(___)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now see what it looks like if you instead group by <code>timepoint</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>table_02 <span class="sc">%&gt;%</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(____) <span class="sc">%&gt;%</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(___)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now try grouping by both <code>timepoint</code> and <code>arm</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>table_02 <span class="sc">%&gt;%</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(__, __) <span class="sc">%&gt;%</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(___)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can perform multiple summaries at once by adding more statements inside the summarise() function, separated by a <code>,</code>. To try this out, calculate the median <code>nugent</code> in addition to the median ph</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>table_02 <span class="sc">%&gt;%</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(____) <span class="sc">%&gt;%</span></span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(___, ___)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>When performing summaries, we often want to know how many observations there are in each group (i.e., we want to count). We can do this with the function <code>n()</code>, which inside summarise() gives us the group size. So, we can count by adding a statement such as <code>count = n()</code> inside `summarise(). Try this out.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>table_02 <span class="sc">%&gt;%</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(____) <span class="sc">%&gt;%</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(___, ___)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="making-tables-wider-or-longer" class="level3">
<h3 class="anchored" data-anchor-id="making-tables-wider-or-longer">Making tables wider or longer</h3>
<p>For efficient data processing, we usually want tables in long form, where each columns is one variable and each row is one observation. However, in some applications, for example when making a table easier to read for humans, a wide format can be preferred. In a wide format, some variables are displayed as column names, and other variables are distributed over multiple columns.</p>
<p>First, make a summary table that shows median <code>ph</code> by <code>arm</code> and <code>time_point</code>, just like you did above, and save it to a variable <code>ph_summary_long</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>ph_summary_long <span class="ot">&lt;-</span> table_02 <span class="sc">%&gt;%</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(___) <span class="sc">%&gt;%</span></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(___)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, try using <code>pivot_wider()</code> to make a column for each arm. Remember, use <code>?pivot_wider</code> if you want help, and try asking google or chatGPT if you get stuck.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>ph_summary_long <span class="sc">%&gt;%</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(____)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>What if you wanted to instead make a column for each time point, and have the arms be different rows?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>ph_summary_long <span class="sc">%&gt;%</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(____)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="combining-datasets-with-joins" class="level3">
<h3 class="anchored" data-anchor-id="combining-datasets-with-joins">Combining datasets with joins</h3>
<p>Finally, we sometimes encounter the situation where we have two data sets that contain different pieces of information about the same subjects or objects, and we need to merge these tables for further analysis. In this situation, we need to perform a join, and there are multiple different types of joins available: left_join(), right_join(), inner_join(), full_join(). These joins all differ in how they handle cases where an observation is present in only one of the two tables but missing in the other.</p>
<p>Our instructional dataset has no missing values, so all types of joins are actually equivalent. Try to join <code>table_01</code> and <code>table_02</code> using <code>left_join()</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="in"># join table_01 and table_02</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="advanced-next-steps" class="level2">
<h2 class="anchored" data-anchor-id="advanced-next-steps">Advanced next steps</h2>
<p>Join together the cytokine data and the metadata and explore the relationships between timepoint, arm, and the different cytokines.</p>
<p>Note, to do this you will need to use <a href="datasets/instructional_dataset/00_sample_ids.csv"><code>00_sample_ids.csv</code></a>, <a href="datasets/instructional_dataset/01_participant_metadata.csv"><code>01_participant_metadata.csv</code></a>, and <a href="datasets/instructional_dataset/03_elisa_cytokines.csv"><code>03_elisa_cytokines.csv</code></a> .</p>
<p>Also look at the relationship of vaginal pH to the different cytokines. Now you’ll need to join the <a href="datasets/instructional_dataset/02_visit_clinical_measurements.csv"><code>02_visit_clinical_measurements.csv</code></a> table as well.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/kwondry\.github\.io\/documentation\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../materials/getting-started/local_R/index.html" class="pagination-link" aria-label="</b> RStudio and recommended packages">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"> RStudio and recommended packages</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../materials/rnaseq/background/index.html" class="pagination-link" aria-label="Background">
        <span class="nav-page-text">Background</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>© 2023. Huge thanks to <a href="https://github.com/andrewpbray">Andrew P. Bray</a> and <a href="https://github.com/clauswilke">Claus O. Wilke</a> for open source code and sharing content.</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>This page is built with <i class="fa-solid fa-heart" aria-label="heart"></i> and <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>




</body></html>